Fix the bug that makes all UI tests fail to start when starting the IDE takes longer

Automated UI tests starts as 2 separated processes (required by the framework we use for UI tests):

**./gradlew clean runIdeForUiTests** - starts the IntelliJ for UI tests
**./gradlew integrationTest** - starts the UI tests

A fixed timeout in the framework we use for our UI tests makes all the UI tests to fail when the IDE takes longer to show the UI visible on the screen when starting the IDE (IDE needs to be downloaded and the process with the UI tests is already running and waiting).

Well we are not able to automate the tests by running the 2 commands at the same time:
**./gradlew clean runIdeForUiTests& ./gradlew integrationTest**


```
java.net.ConnectException: Failed to connect to /127.0.0.1:8082
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:265)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:183)
	at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.java:224)
	at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.java:108)
	at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.java:88)
	at okhttp3.internal.connection.Transmitter.newExchange(Transmitter.java:169)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:41)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:229)
	at okhttp3.RealCall.execute(RealCall.java:81)
	at retrofit2.OkHttpCall.execute(OkHttpCall.java:188)
	at com.intellij.remoterobot.client.IdeRobotClient.findAllByXpath(IdeRobotClient.kt:38)
	at com.intellij.remoterobot.search.Finder.findMany(Finder.kt:41)
	at com.intellij.remoterobot.SearchContext$find$1$1.invoke(SearchContext.kt:55)
	at com.intellij.remoterobot.SearchContext$find$1$1.invoke(SearchContext.kt:15)
	at com.intellij.remoterobot.utils.RepeatUtilsKt$waitFor$1.invoke(RepeatUtils.kt:27)
	at com.intellij.remoterobot.utils.RepeatUtilsKt$waitFor$1.invoke(RepeatUtils.kt)
	at com.intellij.remoterobot.utils.RepeatUtilsKt.repeatInTime(RepeatUtils.kt:57)
	at com.intellij.remoterobot.utils.RepeatUtilsKt.waitFor(RepeatUtils.kt:27)
	at com.intellij.remoterobot.utils.RepeatUtilsKt.waitFor$default(RepeatUtils.kt:24)
	at com.intellij.remoterobot.SearchContext$find$1.invoke(SearchContext.kt:51)
	at com.intellij.remoterobot.SearchContext$find$1.invoke(SearchContext.kt:15)
	at com.intellij.remoterobot.stepsProcessing.StepWorkerKt.step(StepWorker.kt:21)
	at com.intellij.remoterobot.SearchContext$DefaultImpls.find(SearchContext.kt:49)
	at com.intellij.remoterobot.RemoteRobot.find(RemoteRobot.kt:26)
	at com.intellij.remoterobot.SearchContext$DefaultImpls.find(SearchContext.kt:32)
	at com.intellij.remoterobot.RemoteRobot.find(RemoteRobot.kt:26)
	at com.intellij.remoterobot.SearchContext$DefaultImpls.find(SearchContext.kt:26)
	at com.intellij.remoterobot.RemoteRobot.find(RemoteRobot.kt:26)
	at org.jboss.tools.intellij.quarkus.utils.GlobalUtils.lambda$quitIntelliJFromTheWelcomeDialog$9(GlobalUtils.java:134)
	at com.intellij.remoterobot.stepsProcessing.StepWorkerKt$step$1.invoke(StepWorker.kt:14)
	at com.intellij.remoterobot.stepsProcessing.StepWorkerKt$step$1.invoke(StepWorker.kt)
	at com.intellij.remoterobot.stepsProcessing.StepWorkerKt.step(StepWorker.kt:21)
	at com.intellij.remoterobot.stepsProcessing.StepWorkerKt.step(StepWorker.kt:13)
	at org.jboss.tools.intellij.quarkus.utils.GlobalUtils.quitIntelliJFromTheWelcomeDialog(GlobalUtils.java:133)
	at org.jboss.tools.intellij.quarkus.BasicTests.quitTheIde(BasicTests.java:47)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:515)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.lambda$invokeAfterAllMethods$9(ClassTestDescriptor.java:387)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:72)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.lambda$invokeAfterAllMethods$10(ClassTestDescriptor.java:386)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1080)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.invokeAfterAllMethods(ClassTestDescriptor.java:386)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.after(ClassTestDescriptor.java:217)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.after(ClassTestDescriptor.java:74)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:119)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:72)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:119)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:71)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:110)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:72)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:71)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:132)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:412)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at okhttp3.internal.platform.Platform.connectSocket(Platform.java:130)
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:263)
	... 105 more
```

--- a/src/it/java/org/jboss/tools/intellij/quarkus/BasicTests.java
+++ b/src/it/java/org/jboss/tools/intellij/quarkus/BasicTests.java
@@ -32,6 +32,7 @@ public class BasicTests {
 
     @BeforeAll
     public static void connect() throws InterruptedException {
+        GlobalUtils.waitUntilIntelliJStarts();
         robot = new RemoteRobot("http://127.0.0.1:8082");
         for (int i = 0; i < 60; i++) {
             try {

--- a/src/it/java/org/jboss/tools/intellij/quarkus/utils/GlobalUtils.java
+++ b/src/it/java/org/jboss/tools/intellij/quarkus/utils/GlobalUtils.java
@@ -21,6 +21,9 @@ import org.jboss.tools.intellij.quarkus.fixtures.mainIdeWindow.IdeStatusBarFixtu
 
 import java.io.File;
 import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.net.SocketAddress;
 import java.time.Duration;
 
 import static com.intellij.remoterobot.search.locators.Locators.byXpath;
@@ -181,5 +184,25 @@ public class GlobalUtils {
             // the list with accessible name 'Recent Projects' is not available -> 0 links in the 'Welcome to IntelliJ IDEA' dialog
             return 0;
         }
+    }
+
+    public static void waitUntilIntelliJStarts() {
+        waitFor(Duration.ofSeconds(300), Duration.ofSeconds(3), "The IntelliJ Idea did not start in 5 minutes.", () -> isIntelliJUIVisible());
+    }
+
+    private static boolean isIntelliJUIVisible() {
+        return isHostOnIpAndPortAccessible("127.0.0.1", 8082);
+    }
+
+    private static boolean isHostOnIpAndPortAccessible(String ip, int port) {
+        SocketAddress sockaddr = new InetSocketAddress(ip, port);
+        Socket socket = new Socket();
+
+        try {
+            socket.connect(sockaddr, 10000);
+        } catch (IOException IOException) {
+            return false;
+        }
+        return true;
     }
 }

